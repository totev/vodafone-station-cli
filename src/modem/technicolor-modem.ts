import { Log } from '../logger'
import { DocsisChannelType, DocsisStatus, HumanizedDocsis31ChannelStatus, HumanizedDocsisChannelStatus, Modem, normalizeModulation } from './modem'
import { deriveKeyTechnicolor } from './tools/crypto'

export interface TechnicolorBaseResponse{
  error: string | 'ok' | 'error';
  message: string;
  data?: {[key: string]: unknown};
}

export interface TechnicolorConfiguration extends TechnicolorBaseResponse{
    data: {
        LanMode: 'router' | 'modem';
        DeviceMode: 'Ipv4' | 'Ipv4';
        firmwareversion: string;
        internetipv4: string;
        AFTR: string;
        IPAddressRT: string[];
    };
}

export interface TechnicolorSaltResponse  extends TechnicolorBaseResponse{
    salt: string;
    saltwebui: string;
}

export interface TechnicolorTokenResponse  extends TechnicolorBaseResponse{
    token: string;
}

export interface TechnicolorDocsisStatus {
  error: string;
  message: string;
  data: {
    downstream: TechnicolorDocsisChannelStatus[];
    ofdm_downstream: TechnicolorDocsis31ChannelStatus[];
    upstream: TechnicolorDocsisUpstreamChannelStatus[];
    ofdma_upstream: TechnicolorDocsis31UpstreamChannelStatus[];
  };
}

export interface TechnicolorDocsisUpstreamChannelStatus extends Omit<TechnicolorDocsisChannelStatus, 'channelid'|'locked'|'SNR'>{
        channelidup: string;
        CentralFrequency: string;
        RangingStatus: string;
}

export interface TechnicolorDocsis31UpstreamChannelStatus{
    __id: string; // '1',
    channelidup: string; // '9',
    start_frequency: string; // '29.800000 MHz',
    end_frequency: string; // '64.750000 MHz',
    power: string; // '44.0 dBmV',
    CentralFrequency: string; // '46 MHz',
    bandwidth: string; // '35 MHz',
    FFT: string; // 'qpsk',
    ChannelType: string; // 'OFDMA',
    RangingStatus: string;// 'Completed'
}

export interface TechnicolorDocsisChannelStatus {
  __id: string;
  channelid: string;
  CentralFrequency: string; // MHz
  power: string; // dBmV
  SNR: string; // dB
  FFT: string; // modulation
  locked: string;
  ChannelType: DocsisChannelType;
}

export interface TechnicolorDocsis31ChannelStatus {
    __id: string;
    channelid_ofdm: string;
    start_frequency: string; // MHz
    end_frequency: string; // MHz
    CentralFrequency_ofdm: string; // MHz
    bandwidth: string; // MHz
    power_ofdm: string; // dBmV
    SNR_ofdm: string; // dB
    FFT_ofdm:  string; // 'qam256/qam1024'
    locked_ofdm: string;
    ChannelType: DocsisChannelType;
}

export function normalizeChannelStatus(channelStatus: TechnicolorDocsisChannelStatus): HumanizedDocsisChannelStatus {
  return {
    channelId: channelStatus.channelid,
    channelType: channelStatus.ChannelType,
    modulation: normalizeModulation(channelStatus.FFT),
    lockStatus: channelStatus.locked,
    snr: parseFloat(`${channelStatus.SNR ?? 0}`),
    frequency: parseInt(`${channelStatus.CentralFrequency ?? 0}`, 10),
    powerLevel: parseFloat(channelStatus.power),
  }
}
export function normalizeUpstreamChannelStatus(channelStatus: TechnicolorDocsisUpstreamChannelStatus): HumanizedDocsisChannelStatus {
  return {
    channelId: channelStatus.channelidup,
    channelType: channelStatus.ChannelType,
    modulation: normalizeModulation(channelStatus.FFT),
    lockStatus: channelStatus.RangingStatus,
    snr: 0,
    frequency: parseInt(`${channelStatus.CentralFrequency ?? 0}`, 10),
    powerLevel: parseFloat(channelStatus.power),
  }
}
export function normalizeUpstreamOfdmaChannelStatus(channelStatus: TechnicolorDocsis31UpstreamChannelStatus): HumanizedDocsis31ChannelStatus {
  return {
    channelId: channelStatus.channelidup,
    lockStatus: channelStatus.RangingStatus,
    channelType: channelStatus.ChannelType as DocsisChannelType,
    modulation: normalizeModulation(channelStatus.FFT),
    powerLevel: parseFloat(channelStatus.power),
    frequencyStart: parseFloat(channelStatus.start_frequency),
    frequencyEnd: parseFloat(channelStatus.end_frequency),
    snr: 0
  }
}

export function normalizeOfdmChannelStatus(channelStatus: TechnicolorDocsis31ChannelStatus): HumanizedDocsis31ChannelStatus {
  return {
    channelId: channelStatus.channelid_ofdm,
    lockStatus: channelStatus.locked_ofdm,
    channelType: channelStatus.ChannelType,
    modulation: normalizeModulation(channelStatus.FFT_ofdm),
    powerLevel: parseFloat(channelStatus.power_ofdm),
    frequencyStart: parseInt(channelStatus.start_frequency, 10),
    frequencyEnd: parseInt(channelStatus.end_frequency, 10),
    snr: parseFloat(channelStatus.SNR_ofdm)
  }
}

export function normalizeDocsisStatus(channelStatus: TechnicolorDocsisStatus): DocsisStatus {
  return {
    downstream: channelStatus.data.downstream.map(normalizeChannelStatus),
    downstreamOfdm: channelStatus.data.ofdm_downstream.map(normalizeOfdmChannelStatus),
    upstream: channelStatus.data.upstream.map(normalizeUpstreamChannelStatus),
    upstreamOfdma: channelStatus.data.ofdma_upstream.map(normalizeUpstreamOfdmaChannelStatus),
    time: new Date().toISOString()
  }
}

export class Technicolor extends Modem {
  constructor(readonly modemIp: string, readonly logger: Log) {
    super(modemIp, logger)
  }

  async login(password: string): Promise<void> {
    try {
      const {data: salt} = await this.httpClient.post<TechnicolorSaltResponse>('/api/v1/session/login', `username=${Modem.USERNAME}&password=seeksalthash`, {
        headers: {
          'Content-Type': 'application/x-www-form-urlencoded; charset=UTF-8',
          'Referer': `http://${this.modemIp}`,
        }
      })
      this.logger.debug('Salt', salt)

      if (salt.message && salt.message === 'MSG_LOGIN_150') {
        throw new Error('A user is already logged in')
      }

      const derivedKey = deriveKeyTechnicolor(deriveKeyTechnicolor(password, salt.salt), salt.saltwebui)
      this.logger.debug('Derived key', derivedKey)
      const {data: loginResponse} = await this.httpClient.post<TechnicolorBaseResponse>('/api/v1/session/login', `username=${Modem.USERNAME}&password=${derivedKey}`, {
        headers: {
          'Content-Type': 'application/x-www-form-urlencoded; charset=UTF-8',
          'Referer': `http://${this.modemIp}`,
        },
      })
      this.logger.debug('Login status', loginResponse)
      const {data: messageResponse} = await this.httpClient.get<TechnicolorBaseResponse>('/api/v1/session/menu', {
        headers: {
          'Content-Type': 'application/x-www-form-urlencoded; charset=UTF-8',
          'Referer': `http://${this.modemIp}`,
        },
      })
      this.logger.debug('Message status', messageResponse)
    } catch (error) {
      this.logger.warn(`Something went wrong with the login ${error}`)
    }
  }

  async docsis(): Promise<DocsisStatus> {
    const {data: docsisStatus} = await this.httpClient.get('/api/v1/sta_docsis_status', {
      headers: {
        'Content-Type': 'application/x-www-form-urlencoded; charset=UTF-8',
        'Referer': `http://${this.modemIp}`,
      },
    })
    return normalizeDocsisStatus(docsisStatus as TechnicolorDocsisStatus)
  }

  async logout(): Promise<void> {
    this.logger.debug('Logging out...')
    return this.httpClient.post('api/v1/session/logout')
  }

  async restart(): Promise<unknown> {
    const { data: tokenResponse } = await this.httpClient.get<TechnicolorTokenResponse>('api/v1/session/init_page', {
      headers: {
        'Referer': `http://${this.modemIp}`,
      }
    })

    this.logger.debug('Token response: ', tokenResponse)
    const {data: restartResponse} = await this.httpClient.post<TechnicolorBaseResponse>('api/v1/sta_restart',
      'restart=Router%2CWifi%2CVoIP%2CDect%2CMoCA&ui_access=reboot_device', {
        headers: {
          'Content-Type': 'application/x-www-form-urlencoded; charset=UTF-8',
          'X-CSRF-TOKEN': tokenResponse.token,
          'Referer': `http://${this.modemIp}`,
        },
      })

    if (restartResponse?.error === 'error') {
      this.logger.debug(restartResponse)
      throw new Error(`Could not restart router: ${restartResponse.message}`)
    }
    return restartResponse
  }

  async firewall(status: boolean = true): Promise<void> {
    const action = status ? 'on' : 'off'
    const { data: tokenResponse } = await this.httpClient.get<TechnicolorTokenResponse>('api/v1/session/init_page', {
      headers: {
        'Referer': `http://${this.modemIp}`,
      }
    })
    this.logger.debug('Token response: ', tokenResponse)
    const {data: restartResponse} = await this.httpClient.post<TechnicolorBaseResponse>('api/v1/firewall',
      `FirewallLevel=${action}&FirewallLevelV6=${action}`, {
        headers: {
          'Content-Type': 'application/x-www-form-urlencoded; charset=UTF-8',
          'X-CSRF-TOKEN': tokenResponse.token,
          'Referer': `http://${this.modemIp}`,
        },
      })

    if (restartResponse?.error === 'error') {
      this.logger.debug(restartResponse)
      throw new Error(`Could not restart router: ${restartResponse.message}`)
    }
  }
}

