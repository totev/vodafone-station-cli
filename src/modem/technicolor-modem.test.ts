import {normalizeChannelStatus, normalizeDocsisStatus, normalizeOfdmChannelStatus, normalizeUpstreamChannelStatus, normalizeUpstreamOfdmaChannelStatus, TechnicolorDocsisStatus} from './technicolor-modem'
import fixtureDocsisStatus from './__fixtures__/docsisStatus_technicolor.json'

import fixtureDocsis31Status from './__fixtures__/docsisStatus_ofdma_technicolor.json'

test('normalizeChannelStatus with SC-QAM channel', () => {
  const nativeStatus =
    {
      __id: '1',
      channelid: '5',
      CentralFrequency: '602 MHz',
      power: '-4.0 dBmV',
      SNR: '38.8 dB',
      FFT: '256 QAM',
      locked: 'Locked',
      ChannelType: 'SC-QAM'
    } as const
  const status = normalizeChannelStatus(nativeStatus)
  expect(status).toEqual(
    {
      channelId: '5',
      channelType: 'SC-QAM',
      modulation: '256 QAM',
      powerLevel: -4,
      lockStatus: 'Locked',
      snr: 38.8,
      frequency: 602
    },
  )
})

test('normalizeOfdmChannelStatus with OFDM channel', () => {
  const nativeStatus =
    {
      __id: '1',
      channelid_ofdm: '33',
      start_frequency: '151 MHz',
      end_frequency: '324 MHz',
      CentralFrequency_ofdm: '288 MHz',
      bandwidth: '171 MHz',
      power_ofdm: '-3.2 dBmV',
      SNR_ofdm: '39.55 dB',
      FFT_ofdm: 'qam256/qam1024',
      locked_ofdm: 'Locked',
      ChannelType: 'OFDM'
    } as const

  const status = normalizeOfdmChannelStatus(nativeStatus)
  expect(status).toEqual(
    {
      channelId: '33',
      channelType: 'OFDM',
      modulation: 'qam256/qam1024',
      powerLevel: -3.2,
      lockStatus: 'Locked',
      snr: 39.55,
      frequencyStart: 151,
      frequencyEnd: 324
    }
  )
})

test('normalizeUpstreamChannelStatus', () => {
  const nativeStatus =             {
    __id: '1',
    channelidup: '1',
    CentralFrequency: '51.0 MHz',
    power: '49.8 dBmV',
    ChannelType: 'SC-QAM',
    FFT: 'qam64',
    RangingStatus: 'Completed'
  } as const
  expect(normalizeUpstreamChannelStatus(nativeStatus)).toEqual(
    {
      channelId: '1',
      channelType: 'SC-QAM',
      frequency: 51,
      lockStatus: 'Completed',
      modulation: 'qam64',
      powerLevel: 49.8,
      snr: 0,
    },
  )
})

test('normalizeUpstreamOfdmaChannelStatus', () => {
  const nativeStatus =            {
    __id: '1',
    channelidup: '9',
    start_frequency: '29.800000 MHz',
    end_frequency: '64.750000 MHz',
    power: '44.0 dBmV',
    CentralFrequency: '46 MHz',
    bandwidth: '35 MHz',
    FFT: 'qpsk',
    ChannelType: 'OFDMA',
    RangingStatus: 'Completed'
  } as const
  expect(normalizeUpstreamOfdmaChannelStatus(nativeStatus)).toEqual(
    {
      channelId: '9',
      channelType: 'OFDMA',
      frequencyEnd: 64.75,
      frequencyStart: 29.8,
      lockStatus: 'Completed',
      modulation: 'qpsk',
      powerLevel: 44,
      snr: 0,
    },
  )
})

describe('normalizeDocsisStatus', () => {
  test('should work with ofdm in download', () => {
    const {time, ...status} = normalizeDocsisStatus(fixtureDocsisStatus as TechnicolorDocsisStatus)
    expect(status).toMatchSnapshot()
  })
  test('should work with ofdm in download and ofdam in upload', () => {
    const {time, ...status} = normalizeDocsisStatus(fixtureDocsis31Status as TechnicolorDocsisStatus)
    expect(status).toMatchSnapshot()
  })
})
