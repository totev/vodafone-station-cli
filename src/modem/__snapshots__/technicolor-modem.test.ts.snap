// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`normalizeDocsisStatus should work with ofdm in download 1`] = `
{
  "downstream": [
    {
      "channelId": "5",
      "channelType": "SC-QAM",
      "frequency": 602,
      "lockStatus": "Locked",
      "modulation": "256QAM",
      "powerLevel": -4,
      "snr": 38.8,
    },
    {
      "channelId": "1",
      "channelType": "SC-QAM",
      "frequency": 114,
      "lockStatus": "Locked",
      "modulation": "256QAM",
      "powerLevel": -5.4,
      "snr": 36.5,
    },
    {
      "channelId": "2",
      "channelType": "SC-QAM",
      "frequency": 130,
      "lockStatus": "Locked",
      "modulation": "256QAM",
      "powerLevel": -5.4,
      "snr": 36.8,
    },
    {
      "channelId": "3",
      "channelType": "SC-QAM",
      "frequency": 138,
      "lockStatus": "Locked",
      "modulation": "256QAM",
      "powerLevel": -5.5,
      "snr": 36.9,
    },
    {
      "channelId": "4",
      "channelType": "SC-QAM",
      "frequency": 146,
      "lockStatus": "Locked",
      "modulation": "256QAM",
      "powerLevel": -5.2,
      "snr": 37.2,
    },
    {
      "channelId": "6",
      "channelType": "SC-QAM",
      "frequency": 618,
      "lockStatus": "Locked",
      "modulation": "256QAM",
      "powerLevel": -3.8,
      "snr": 38.9,
    },
    {
      "channelId": "7",
      "channelType": "SC-QAM",
      "frequency": 626,
      "lockStatus": "Locked",
      "modulation": "256QAM",
      "powerLevel": -3.9,
      "snr": 38.9,
    },
    {
      "channelId": "8",
      "channelType": "SC-QAM",
      "frequency": 642,
      "lockStatus": "Locked",
      "modulation": "256QAM",
      "powerLevel": -4.9,
      "snr": 38.5,
    },
    {
      "channelId": "9",
      "channelType": "SC-QAM",
      "frequency": 650,
      "lockStatus": "Locked",
      "modulation": "256QAM",
      "powerLevel": -4.8,
      "snr": 38.4,
    },
    {
      "channelId": "10",
      "channelType": "SC-QAM",
      "frequency": 658,
      "lockStatus": "Locked",
      "modulation": "256QAM",
      "powerLevel": -4.1,
      "snr": 39.1,
    },
    {
      "channelId": "11",
      "channelType": "SC-QAM",
      "frequency": 666,
      "lockStatus": "Locked",
      "modulation": "256QAM",
      "powerLevel": -4.1,
      "snr": 38.7,
    },
    {
      "channelId": "12",
      "channelType": "SC-QAM",
      "frequency": 674,
      "lockStatus": "Locked",
      "modulation": "256QAM",
      "powerLevel": -3.1,
      "snr": 39,
    },
    {
      "channelId": "13",
      "channelType": "SC-QAM",
      "frequency": 682,
      "lockStatus": "Locked",
      "modulation": "256QAM",
      "powerLevel": -2.8,
      "snr": 39.7,
    },
    {
      "channelId": "14",
      "channelType": "SC-QAM",
      "frequency": 690,
      "lockStatus": "Locked",
      "modulation": "256QAM",
      "powerLevel": -1.8,
      "snr": 40.4,
    },
    {
      "channelId": "15",
      "channelType": "SC-QAM",
      "frequency": 698,
      "lockStatus": "Locked",
      "modulation": "64QAM",
      "powerLevel": -7.7,
      "snr": 34.6,
    },
    {
      "channelId": "16",
      "channelType": "SC-QAM",
      "frequency": 706,
      "lockStatus": "Locked",
      "modulation": "64QAM",
      "powerLevel": -6.8,
      "snr": 35.4,
    },
    {
      "channelId": "17",
      "channelType": "SC-QAM",
      "frequency": 714,
      "lockStatus": "Locked",
      "modulation": "64QAM",
      "powerLevel": -7.1,
      "snr": 35,
    },
    {
      "channelId": "18",
      "channelType": "SC-QAM",
      "frequency": 722,
      "lockStatus": "Locked",
      "modulation": "64QAM",
      "powerLevel": -7.2,
      "snr": 35,
    },
    {
      "channelId": "19",
      "channelType": "SC-QAM",
      "frequency": 730,
      "lockStatus": "Locked",
      "modulation": "64QAM",
      "powerLevel": -6.9,
      "snr": 35.1,
    },
    {
      "channelId": "20",
      "channelType": "SC-QAM",
      "frequency": 738,
      "lockStatus": "Locked",
      "modulation": "64QAM",
      "powerLevel": -5.9,
      "snr": 35.7,
    },
    {
      "channelId": "21",
      "channelType": "SC-QAM",
      "frequency": 746,
      "lockStatus": "Locked",
      "modulation": "64QAM",
      "powerLevel": -5.6,
      "snr": 35.8,
    },
    {
      "channelId": "22",
      "channelType": "SC-QAM",
      "frequency": 754,
      "lockStatus": "Locked",
      "modulation": "64QAM",
      "powerLevel": -6.7,
      "snr": 35,
    },
    {
      "channelId": "23",
      "channelType": "SC-QAM",
      "frequency": 762,
      "lockStatus": "Locked",
      "modulation": "64QAM",
      "powerLevel": -6.4,
      "snr": 35.1,
    },
    {
      "channelId": "24",
      "channelType": "SC-QAM",
      "frequency": 770,
      "lockStatus": "Locked",
      "modulation": "64QAM",
      "powerLevel": -6.3,
      "snr": 35.1,
    },
    {
      "channelId": "25",
      "channelType": "SC-QAM",
      "frequency": 778,
      "lockStatus": "Locked",
      "modulation": "64QAM",
      "powerLevel": -6,
      "snr": 35.7,
    },
    {
      "channelId": "26",
      "channelType": "SC-QAM",
      "frequency": 786,
      "lockStatus": "Locked",
      "modulation": "64QAM",
      "powerLevel": -6.2,
      "snr": 35.2,
    },
    {
      "channelId": "27",
      "channelType": "SC-QAM",
      "frequency": 794,
      "lockStatus": "Locked",
      "modulation": "64QAM",
      "powerLevel": -6.5,
      "snr": 34.8,
    },
    {
      "channelId": "28",
      "channelType": "SC-QAM",
      "frequency": 802,
      "lockStatus": "Locked",
      "modulation": "64QAM",
      "powerLevel": -7.1,
      "snr": 34.7,
    },
    {
      "channelId": "29",
      "channelType": "SC-QAM",
      "frequency": 810,
      "lockStatus": "Locked",
      "modulation": "64QAM",
      "powerLevel": -6.9,
      "snr": 34.9,
    },
    {
      "channelId": "30",
      "channelType": "SC-QAM",
      "frequency": 818,
      "lockStatus": "Locked",
      "modulation": "64QAM",
      "powerLevel": -6.7,
      "snr": 35,
    },
    {
      "channelId": "31",
      "channelType": "SC-QAM",
      "frequency": 826,
      "lockStatus": "Locked",
      "modulation": "64QAM",
      "powerLevel": -7.2,
      "snr": 35,
    },
    {
      "channelId": "32",
      "channelType": "SC-QAM",
      "frequency": 834,
      "lockStatus": "Locked",
      "modulation": "64QAM",
      "powerLevel": -7.8,
      "snr": 34.2,
    },
  ],
  "downstreamOfdm": [
    {
      "channelId": "33",
      "channelType": "OFDM",
      "frequencyEnd": 324,
      "frequencyStart": 151,
      "lockStatus": "Locked",
      "modulation": "QAM256",
      "powerLevel": -3,
      "snr": 39.55,
    },
  ],
  "upstream": [
    {
      "channelId": "1",
      "channelType": "SC-QAM",
      "frequency": 51,
      "lockStatus": "Completed",
      "modulation": "QAM64",
      "powerLevel": 49.8,
      "snr": 0,
    },
    {
      "channelId": "2",
      "channelType": "SC-QAM",
      "frequency": 44,
      "lockStatus": "Completed",
      "modulation": "QAM64",
      "powerLevel": 51.8,
      "snr": 0,
    },
    {
      "channelId": "3",
      "channelType": "SC-QAM",
      "frequency": 37,
      "lockStatus": "Completed",
      "modulation": "QAM64",
      "powerLevel": 51.8,
      "snr": 0,
    },
    {
      "channelId": "4",
      "channelType": "SC-QAM",
      "frequency": 30,
      "lockStatus": "Completed",
      "modulation": "QAM32",
      "powerLevel": 51.8,
      "snr": 0,
    },
  ],
  "upstreamOfdma": [],
}
`;

exports[`normalizeDocsisStatus should work with ofdm in download and ofdam in upload 1`] = `
{
  "downstream": [
    {
      "channelId": "5",
      "channelType": "SC-QAM",
      "frequency": 602,
      "lockStatus": "Locked",
      "modulation": "256QAM",
      "powerLevel": -4.3,
      "snr": 38.9,
    },
    {
      "channelId": "1",
      "channelType": "SC-QAM",
      "frequency": 114,
      "lockStatus": "Locked",
      "modulation": "256QAM",
      "powerLevel": -5.8,
      "snr": 36.4,
    },
    {
      "channelId": "2",
      "channelType": "SC-QAM",
      "frequency": 130,
      "lockStatus": "Locked",
      "modulation": "256QAM",
      "powerLevel": -5.8,
      "snr": 36.8,
    },
    {
      "channelId": "3",
      "channelType": "SC-QAM",
      "frequency": 138,
      "lockStatus": "Locked",
      "modulation": "256QAM",
      "powerLevel": -5.9,
      "snr": 36.9,
    },
    {
      "channelId": "4",
      "channelType": "SC-QAM",
      "frequency": 146,
      "lockStatus": "Locked",
      "modulation": "256QAM",
      "powerLevel": -5.6,
      "snr": 37.2,
    },
    {
      "channelId": "6",
      "channelType": "SC-QAM",
      "frequency": 618,
      "lockStatus": "Locked",
      "modulation": "256QAM",
      "powerLevel": -3.7,
      "snr": 39.1,
    },
    {
      "channelId": "7",
      "channelType": "SC-QAM",
      "frequency": 626,
      "lockStatus": "Locked",
      "modulation": "256QAM",
      "powerLevel": -3.4,
      "snr": 39.5,
    },
    {
      "channelId": "8",
      "channelType": "SC-QAM",
      "frequency": 642,
      "lockStatus": "Locked",
      "modulation": "256QAM",
      "powerLevel": -3.1,
      "snr": 39.8,
    },
    {
      "channelId": "9",
      "channelType": "SC-QAM",
      "frequency": 650,
      "lockStatus": "Locked",
      "modulation": "256QAM",
      "powerLevel": -2.7,
      "snr": 40,
    },
    {
      "channelId": "10",
      "channelType": "SC-QAM",
      "frequency": 658,
      "lockStatus": "Locked",
      "modulation": "256QAM",
      "powerLevel": -2.5,
      "snr": 40,
    },
    {
      "channelId": "11",
      "channelType": "SC-QAM",
      "frequency": 666,
      "lockStatus": "Locked",
      "modulation": "256QAM",
      "powerLevel": -2.9,
      "snr": 39.6,
    },
    {
      "channelId": "12",
      "channelType": "SC-QAM",
      "frequency": 674,
      "lockStatus": "Locked",
      "modulation": "256QAM",
      "powerLevel": -2.6,
      "snr": 39.6,
    },
    {
      "channelId": "13",
      "channelType": "SC-QAM",
      "frequency": 682,
      "lockStatus": "Locked",
      "modulation": "256QAM",
      "powerLevel": -2.6,
      "snr": 40.1,
    },
    {
      "channelId": "14",
      "channelType": "SC-QAM",
      "frequency": 690,
      "lockStatus": "Locked",
      "modulation": "256QAM",
      "powerLevel": -1.7,
      "snr": 40.6,
    },
    {
      "channelId": "15",
      "channelType": "SC-QAM",
      "frequency": 698,
      "lockStatus": "Locked",
      "modulation": "64QAM",
      "powerLevel": -7.9,
      "snr": 34.7,
    },
    {
      "channelId": "16",
      "channelType": "SC-QAM",
      "frequency": 706,
      "lockStatus": "Locked",
      "modulation": "64QAM",
      "powerLevel": -7.1,
      "snr": 35.4,
    },
    {
      "channelId": "17",
      "channelType": "SC-QAM",
      "frequency": 714,
      "lockStatus": "Locked",
      "modulation": "64QAM",
      "powerLevel": -7.5,
      "snr": 34.8,
    },
    {
      "channelId": "18",
      "channelType": "SC-QAM",
      "frequency": 722,
      "lockStatus": "Locked",
      "modulation": "64QAM",
      "powerLevel": -7.6,
      "snr": 35,
    },
    {
      "channelId": "19",
      "channelType": "SC-QAM",
      "frequency": 730,
      "lockStatus": "Locked",
      "modulation": "64QAM",
      "powerLevel": -7.2,
      "snr": 35.1,
    },
    {
      "channelId": "20",
      "channelType": "SC-QAM",
      "frequency": 738,
      "lockStatus": "Locked",
      "modulation": "64QAM",
      "powerLevel": -6.3,
      "snr": 35.8,
    },
    {
      "channelId": "21",
      "channelType": "SC-QAM",
      "frequency": 746,
      "lockStatus": "Locked",
      "modulation": "64QAM",
      "powerLevel": -6.2,
      "snr": 35.7,
    },
    {
      "channelId": "22",
      "channelType": "SC-QAM",
      "frequency": 754,
      "lockStatus": "Locked",
      "modulation": "64QAM",
      "powerLevel": -7.2,
      "snr": 35,
    },
    {
      "channelId": "23",
      "channelType": "SC-QAM",
      "frequency": 762,
      "lockStatus": "Locked",
      "modulation": "64QAM",
      "powerLevel": -7,
      "snr": 35,
    },
    {
      "channelId": "24",
      "channelType": "SC-QAM",
      "frequency": 770,
      "lockStatus": "Locked",
      "modulation": "64QAM",
      "powerLevel": -6.7,
      "snr": 35,
    },
    {
      "channelId": "25",
      "channelType": "SC-QAM",
      "frequency": 778,
      "lockStatus": "Locked",
      "modulation": "64QAM",
      "powerLevel": -6.3,
      "snr": 35.8,
    },
    {
      "channelId": "26",
      "channelType": "SC-QAM",
      "frequency": 786,
      "lockStatus": "Locked",
      "modulation": "64QAM",
      "powerLevel": -6.4,
      "snr": 35.3,
    },
    {
      "channelId": "27",
      "channelType": "SC-QAM",
      "frequency": 794,
      "lockStatus": "Locked",
      "modulation": "64QAM",
      "powerLevel": -6.5,
      "snr": 35.3,
    },
    {
      "channelId": "28",
      "channelType": "SC-QAM",
      "frequency": 802,
      "lockStatus": "Locked",
      "modulation": "64QAM",
      "powerLevel": -6.8,
      "snr": 35.2,
    },
    {
      "channelId": "29",
      "channelType": "SC-QAM",
      "frequency": 810,
      "lockStatus": "Locked",
      "modulation": "64QAM",
      "powerLevel": -7,
      "snr": 35,
    },
    {
      "channelId": "30",
      "channelType": "SC-QAM",
      "frequency": 818,
      "lockStatus": "Locked",
      "modulation": "64QAM",
      "powerLevel": -6.4,
      "snr": 35.3,
    },
    {
      "channelId": "31",
      "channelType": "SC-QAM",
      "frequency": 826,
      "lockStatus": "Locked",
      "modulation": "64QAM",
      "powerLevel": -7,
      "snr": 35.2,
    },
    {
      "channelId": "32",
      "channelType": "SC-QAM",
      "frequency": 834,
      "lockStatus": "Locked",
      "modulation": "64QAM",
      "powerLevel": -7.9,
      "snr": 34.4,
    },
  ],
  "downstreamOfdm": [
    {
      "channelId": "33",
      "channelType": "OFDM",
      "frequencyEnd": 324,
      "frequencyStart": 151,
      "lockStatus": "Locked",
      "modulation": "QAM256",
      "powerLevel": -3.6,
      "snr": 39.39,
    },
  ],
  "upstream": [
    {
      "channelId": "1",
      "channelType": "SC-QAM",
      "frequency": 51,
      "lockStatus": "Completed",
      "modulation": "QAM64",
      "powerLevel": 51.8,
      "snr": 0,
    },
    {
      "channelId": "2",
      "channelType": "SC-QAM",
      "frequency": 44,
      "lockStatus": "Completed",
      "modulation": "QAM64",
      "powerLevel": 51.8,
      "snr": 0,
    },
    {
      "channelId": "3",
      "channelType": "SC-QAM",
      "frequency": 37,
      "lockStatus": "Completed",
      "modulation": "QAM64",
      "powerLevel": 51.8,
      "snr": 0,
    },
    {
      "channelId": "4",
      "channelType": "SC-QAM",
      "frequency": 30,
      "lockStatus": "Completed",
      "modulation": "QAM32",
      "powerLevel": 51.8,
      "snr": 0,
    },
  ],
  "upstreamOfdma": [
    {
      "channelId": "9",
      "channelType": "OFDMA",
      "frequencyEnd": 64.75,
      "frequencyStart": 29.8,
      "lockStatus": "Completed",
      "modulation": "QPSK",
      "powerLevel": 44,
      "snr": 0,
    },
  ],
}
`;
